proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,221 @@
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />` to remove any `auditSources` inherited from parent *nuget.config* files.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+Omitting `<clear />` will not use `packageSources` in addition to `auditSources`.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <!-- Clear to ensure package sources not inherited from other config files -->
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <!-- Clear to ensure audit sources not inherited from other config files -->
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
```

Comment:
I'm trying to understand the "need" here. If a feed does not implement the `VulnerabilityInfo` but it is set as an audit source, what happens? Will nuget error out? Warn? Silently act similar to a package source with no `VulnerabilityInfo`?

I'm thinking a user that is unaware of the protocol will naively put their ADO or other package source also as an audit source optimistically hoping for vulnerability info to start showing up. Can we guide them along the way to a successful config?

Reply:
I'd err on silently doing nothing. But in the grand scheme, we should validate this and empower the user to rightful resolution.

Reply:
IMO, customers won't be able to "rely" on NuGetAudit, unless there's a way to warn or error when audit doesn't run (for example, no VDB could be obtained from any package source). If they expect audit to run as part of their pipeline, they should have a way to fail the pipeline if it doesn't run. And considering the only purpose of `auditSources` are to define sources used to get a VDB, I think it makes extra sense to warn if the configured audit source doesn't have a VDB.

However Joel's the first person who has given me feedback that appears to agree with my position, but since it's still only 1 person, so perhaps we can wait for more customer feedback.

Reply:
I think it wouldn't be the worst thing even error when source configured as an "auditSource" doesn't have a VDB.

Reply:
The same question occurred to me before I read Joel's question. 

Let's ignore for a moment that we may want to update the wordings of NU1900 and NU1905 with this feature or introduce new audit source-specific versions.

Currently, if I enable NuGetAudit, and no package source provides a VDB, then I expect [NU1905](https://learn.microsoft.com/nuget/concepts/auditing-packages#warning-codes).  The V3 `VulnerabilityResource` is an optional resource that each package source can implement (or not), and for backward compatibility reasons we don't require any package source to provide the resource.

The assumption for audit sources is that they do provide a VDB.  In fact, that is the _sole purpose_ of an audit source.

With this feature implemented, if I enable NuGetAudit and specify two audit sources, and one audit source provides a VDB but the other does not, I feel like that should still raise a warning.

Each audit source that does not provide a VDB should have its own warning.